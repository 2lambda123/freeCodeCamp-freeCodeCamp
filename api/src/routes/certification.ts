import { type FastifyPluginCallbackTypebox } from '@fastify/type-provider-typebox';
import { find } from 'lodash';
import { CompletedChallenge } from '@prisma/client';

import { schemas } from '../schemas';
import {
  certSlugTypeMap,
  certTypeIdMap,
  certTypeTitleMap,
  completionHours,
  oldDataVizId,
  certIds
} from '../../../shared/config/certification-settings';
import { formatCertificationValidation } from '../utils/error-formatting';

const {
  legacyInfosecQaId,
  respWebDesignId,
  frontEndDevLibsId,
  jsAlgoDataStructId,
  dataVis2018Id,
  apisMicroservicesId
} = certIds;

function getFallbackFullStackDate(
  completedChallenges: CompletedChallenge[],
  completedDate: number
) {
  const chalIds = [
    respWebDesignId,
    jsAlgoDataStructId,
    frontEndDevLibsId,
    dataVis2018Id,
    apisMicroservicesId,
    legacyInfosecQaId
  ];

  const latestCertDate = completedChallenges
    .filter(chal => chalIds.includes(chal.id))
    .sort((a, b) => b.completedDate - a.completedDate)[0]?.completedDate;

  return latestCertDate ? latestCertDate : completedDate;
}

/**
 * Plugin for the certification submission endpoints.
 *
 * @param fastify The Fastify instance.
 * @param _options Options passed to the plugin via `fastify.register(plugin, options)`.
 * @param done The callback to signal that the plugin is ready.
 */
export const certRoutes: FastifyPluginCallbackTypebox = (
  fastify,
  _options,
  done
) => {
  fastify.get(
    '/certificate/:username/:certSlug',
    {
      schema: schemas.certSlug,
      errorHandler(error, request, reply) {
        if (error.validation) {
          void reply.code(400);
          return formatCertificationValidation(error.validation);
        } else {
          fastify.errorHandler(error, request, reply);
        }
      }
    },
    async (req, reply) => {
      try {
        let { username } = req.params;
        const { certSlug } = req.params;
        username = username.toLowerCase();

        const certType = certSlugTypeMap[certSlug];
        const certId = certTypeIdMap[certType];
        const certTitle = certTypeTitleMap[certType];
        const completionTime = completionHours[certType] || 300;
        const user = await fastify.prisma.user.findFirst({
          where: { username },
          select: {
            isBanned: true,
            isCheater: true,
            isFrontEndCert: true,
            isBackEndCert: true,
            isFullStackCert: true,
            isRespWebDesignCert: true,
            isFrontEndLibsCert: true,
            isJsAlgoDataStructCert: true,
            isDataVisCert: true,
            is2018DataVisCert: true,
            isApisMicroservicesCert: true,
            isInfosecQaCert: true,
            isQaCertV7: true,
            isInfosecCertV7: true,
            isSciCompPyCertV7: true,
            isDataAnalysisPyCertV7: true,
            isMachineLearningPyCertV7: true,
            isRelationalDatabaseCertV8: true,
            isCollegeAlgebraPyCertV8: true,
            // isFoundationalCSharpCertV8: true,
            isHonest: true,
            username: true,
            name: true,
            completedChallenges: true,
            profileUI: true
          }
        });
        if (user === null) {
          return {
            type: 'info',
            message: 'flash.username-not-found',
            variables: { username: username }
          } as const;
        }
        if (!user.name) {
          return {
            type: 'info',
            message: 'flash.add-name'
          } as const;
        }

        if (user.isCheater || user.isBanned) {
          return {
            type: 'info',
            message: 'flash.not-eligible'
          } as const;
        }
        if (!user.isHonest) {
          return {
            type: 'info',
            message: 'flash.not-honest',
            variables: { username: username }
          } as const;
        }

        if (user.profileUI?.isLocked) {
          return {
            type: 'info',
            message: 'flash.profile-private',
            variables: { username: username }
          } as const;
        }
        if (!user.profileUI?.showCerts) {
          return {
            type: 'info',
            message: 'flash.certs-private',
            variables: { username: username }
          } as const;
        }
        if (!user.profileUI?.showTimeLine) {
          return {
            type: 'info',
            message: 'flash.timeline-private',
            variables: { username: username }
          } as const;
        }

        if (user[certType]) {
          const { completedChallenges } = user;
          const certChallenge = find(
            completedChallenges,
            ({ id }) => certId === id
          );
          let { completedDate } = certChallenge || {};

          if (!completedDate) completedDate = Date.now();

          // the challenge id has been rotated for isDataVisCert
          if (certType === 'isDataVisCert' && !certChallenge) {
            const oldDataVisIdChall = find(
              completedChallenges,
              ({ id }) => oldDataVizId === id
            );

            if (oldDataVisIdChall) {
              completedDate = oldDataVisIdChall.completedDate || completedDate;
            }
          }

          // if fullcert is not found, return the latest completedDate
          if (certType === 'isFullStackCert' && !certChallenge) {
            completedDate = getFallbackFullStackDate(
              completedChallenges,
              completedDate
            );
          }

          // if fullcert is not found, return the latest completedDate
          if (certType === 'isFullStackCert' && !certChallenge) {
            completedDate = getFallbackFullStackDate(
              completedChallenges,
              completedDate
            );
          }

          const { username, name } = user;

          if (!user.profileUI.showName) {
            return {
              certSlug,
              certTitle,
              username,
              date: completedDate,
              completionTime
            };
          }

          return {
            certSlug,
            certTitle,
            username,
            name,
            date: completedDate,
            completionTime
          };
        }
      } catch (err) {
        fastify.log.error(err);
        void reply.code(400);
        return { message: 'flash.user-not-certified', type: 'info' } as const;
      }
    }
  );

  done();
};
