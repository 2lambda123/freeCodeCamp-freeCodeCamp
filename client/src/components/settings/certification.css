.certification-settings tr {
  height: 57px;
}

.certification-settings .project-title > a {
  line-height: 40px;
}

.modal {
  overflow-y: auto;
}

.legacy-full-stack-cert-list {
  margin-top: 0;
  margin-bottom: 12.5px;
  padding-inline-start: 40px;
  list-style-type: disc;
}
