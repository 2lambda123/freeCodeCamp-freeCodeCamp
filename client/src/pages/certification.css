@font-face {
  font-family: 'Sax Mono';
  src: url('/fonts/saxmono.ttf') format('truetype');
}

.certification-namespace * {
  margin: 0;
  padding: 0;
}
.certification-namespace h1 {
  margin: 12px 0;
}
.certification-namespace.container {
  max-width: 1500px;
  width: 100%;
  padding: 30px;
  border: darkgreen 15px solid;
  border-radius: 3px;
}

.certification-namespace .row {
  margin: 0;
}

.certificate-outer-wrapper {
  margin-top: 1em;
}

@media (max-width: 1515px) {
  .certificate-outer-wrapper {
    margin-top: 0;
  }
}

.certification-namespace.certificate-inner-wrapper {
  font-family: 'Sax Mono', monospace;
}

.certification-namespace header {
  width: 100%;
  min-height: 140px;
  background-color: darkgreen !important;
  position: relative;
}

.certification-namespace .logo {
  display: flex;
  align-items: center;
  height: 140px;
  margin-left: 100px;
}

.certification-namespace .logo img {
  max-width: 500px !important;
  width: 100%;
}

.certification-namespace .issue-date {
  line-height: 140px;
  font-size: 20px;
  text-align: right;
  margin-right: 100px;
  color: #fff !important;
}

.certification-namespace .issue-date strong {
  color: #fff !important;
}

.certification-namespace .information {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 380px;
  text-align: center;
  background-color: #efefef !important;
}

.certification-namespace p {
  margin: 0;
}

.certification-namespace h3 {
  font-size: 30px;
}

.certification-namespace h4 {
  margin-top: 25px;
  font-size: 20px;
}

.certification-namespace h1 {
  font-size: 40px;
}

.certification-namespace h1 strong {
  color: #006400;
}

.certification-namespace .signatures {
  text-align: center;
  margin: 0 auto;
  background-color: #efefef !important;
}

.certification-namespace .signatures img {
  max-width: 300px !important;
  width: 100%;
  margin: 0 auto;
}

.certification-namespace .signatures p {
  font-size: 18px;
  padding-top: 10px;
}

.certification-namespace .verify {
  padding: 30px 0;
  font-size: 15px;
  text-align: center;
  word-wrap: break-word;
  background-color: #efefef !important;
}

/*adds vertical alignment when the height is bigger than required to display everything*/
@media (min-height: 700px) {
  .certification-namespace body {
    display: flex;
    flex-direction: column;
    justify-content: center;
    height: 100vh;
  }
}

/*mobile media queries*/
@media (max-width: 991px) {
  .certification-namespace header {
    min-height: 160px;
  }

  .certification-namespace .logo {
    margin-left: 0;
    padding: 20px;
    justify-content: center;
    height: 80px;
  }

  .certification-namespace .logo img {
    margin-top: 20px;
  }

  .certification-namespace .issue-date {
    padding: 20px 0;
    margin-right: 0;
    text-align: center;
    line-height: 0px;
    word-wrap: break-word;
  }

  .certification-namespace .issue-date strong {
    display: block;
    margin-top: 15px;
    line-height: 25px;
  }

  .certification-namespace .information-container {
    margin: 15px;
    text-align: center;
    word-break: break-word;
  }

  .certification-namespace h3 {
    font-size: 25px;
  }

  .certification-namespace h1 {
    font-size: 28px;
  }
}

@media (max-width: 675px) {
  .certification-namespaces .issue-date {
    padding: 0;
    border: 0;
  }

  .certification-namespace h3 {
    font-size: 15px;
  }

  .certification-namespace h1 {
    font-size: 17px;
  }

  .certification-namespace h4 {
    font-size: 15px;
    margin-top: 20px;
  }
}

/* print styles */

@media print {
  @page {
    margin: 10px;
  }

  * {
    color-adjust: exact !important;
  }

  .certification-namespace .container {
    padding: 25px;
  }

  .certification-namespace .information {
    min-height: 0;
    padding: 25px 0;
  }

  .certification-namespace .verify {
    padding: 25px 0 15px;
  }

  .certificate-inner-wrapper .certification-namespace .container {
    max-height: 772px;
  }
}
