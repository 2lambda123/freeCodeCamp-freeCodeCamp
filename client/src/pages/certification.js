import React from 'react';
import { Router } from '@reach/router';
import { useStaticQuery, graphql, withPrefix } from 'gatsby';

import RedirectHome from '../components/RedirectHome';
import ShowCertification from '../client-only-routes/ShowCertification';

import './certification.css';

const Certification = () => {
  const certMap = useStaticQuery(graphql`
    query {
      allCertificateNode {
        nodes {
          title
          block
          curriculumVersion
        }
      }
    }
  `).allCertificateNode.nodes;
  const validCertNames = certMap.map(cert =>
    cert.block.replace(
      /-certificate/,
      cert.curriculumVersion === 7 ? '-v7' : ''
    )
  );

  return (
    <Router>
      <ShowCertification
        path={withPrefix('/certification/:username/:certName')}
        validCertNames={validCertNames}
      />
      <RedirectHome default={true} />
    </Router>
  );
};

export default Certification;
