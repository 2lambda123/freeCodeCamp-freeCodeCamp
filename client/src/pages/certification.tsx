import React from 'react';
import { Router } from '@reach/router';
import { withPrefix } from 'gatsby';

import RedirectHome from '../components/RedirectHome';
import ShowCertification from '../client-only-routes/show-certification';

import './certification.css';

function Certification(): JSX.Element {
  return (
    <Router>
      <ShowCertification
        path={withPrefix('/certification/:username/:certSlug')}
      />
      {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
      {/* @ts-ignore */}
      <RedirectHome default={true} />
    </Router>
  );
}

export default Certification;
