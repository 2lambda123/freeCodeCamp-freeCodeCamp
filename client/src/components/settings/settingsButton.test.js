import renderer from 'react-test-renderer';

import {
  legacyProjectMap,
  projectMap
} from '../../resources/cert-and-project-map';
import { CertificationSettings } from './certification';

it('should check legacy certification button consistency', () => {
  const certificationSettings = new CertificationSettings();

  certificationSettings.props = { t: t => t };

  const { renderCertifications } = certificationSettings;

  const legacyCertifications = Object.keys(legacyProjectMap);

  const tree = renderer.create(
    legacyCertifications.map(certName =>
      renderCertifications(certName, legacyProjectMap)
    )
  );

  expect(tree).toMatchSnapshot();
});

// const first = jest.fn(arr => arr[0]);

it('should check certification button consistency', () => {
  const certificationSettings = new CertificationSettings();

  certificationSettings.props = { t: t => t };

  const { renderCertifications } = certificationSettings;

  const certifications = Object.keys(projectMap);

  const tree = renderer.create(
    certifications.map(certName =>
      renderCertifications(certName, certifications)
    )
  );

  expect(tree).toMatchSnapshot();
});
