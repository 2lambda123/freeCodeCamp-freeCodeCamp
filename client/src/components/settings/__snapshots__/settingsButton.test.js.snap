// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should check legacy certification button consistency 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-sm-8 col-sm-offset-2 col-xs-12"
    >
      <div
        className="spacer"
        style={
          Object {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <h3
        className="text-center"
        id="cert-legacy-front-end"
      >
        Legacy Front End
      </h3>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              settings.labels.project-name
            </th>
            <th>
              settings.labels.solution
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/responsive-web-design/responsive-web-design-projects/build-a-personal-portfolio-webpage"
              >
                Build a Personal Portfolio Webpage
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/front-end-development-libraries/front-end-development-libraries-projects/build-a-random-quote-machine"
              >
                Build a Random Quote Machine
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/front-end-development-libraries/front-end-development-libraries-projects/build-a-25--5-clock"
              >
                Build a 25 + 5 Clock
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/front-end-development-libraries/front-end-development-libraries-projects/build-a-javascript-calculator"
              >
                Build a JavaScript Calculator
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/show-the-local-weather"
              >
                Show the Local Weather
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/use-the-twitch-json-api"
              >
                Use the TwitchTV JSON API
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/responsive-web-design/responsive-web-design-projects/build-a-tribute-page"
              >
                Build a Tribute Page
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-wikipedia-viewer"
              >
                Build a Wikipedia Viewer
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-tic-tac-toe-game"
              >
                Build a Tic Tac Toe Game
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-simon-game"
              >
                Build a Simon Game
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr>
            <td
              colSpan={2}
            >
              <a
                className="col-xs-12 btn btn-primary btn-block"
                data-cy="btn-for-legacy-front-end"
                href="/certification/undefined/legacy-front-end"
                onClick={[Function]}
                onKeyDown={[Function]}
              >
                buttons.claim-cert
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-sm-8 col-sm-offset-2 col-xs-12"
    >
      <div
        className="spacer"
        style={
          Object {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <h3
        className="text-center"
        id="cert-legacy-back-end"
      >
        Legacy Back End
      </h3>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              settings.labels.project-name
            </th>
            <th>
              settings.labels.solution
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/back-end-development-and-apis/back-end-development-and-apis-projects/timestamp-microservice"
              >
                Timestamp Microservice
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/back-end-development-and-apis/back-end-development-and-apis-projects/request-header-parser-microservice"
              >
                Request Header Parser Microservice
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/back-end-development-and-apis/back-end-development-and-apis-projects/url-shortener-microservice"
              >
                URL Shortener Microservice
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-an-image-search-abstraction-layer"
              >
                Image Search Abstraction Layer
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/back-end-development-and-apis/back-end-development-and-apis-projects/file-metadata-microservice"
              >
                File Metadata Microservice
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-voting-app"
              >
                Build a Voting App
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-nightlife-coordination-app"
              >
                Build a Nightlife Coordination App
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/chart-the-stock-market"
              >
                Chart the Stock Market
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/manage-a-book-trading-club"
              >
                Manage a Book Trading Club
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-pinterest-clone"
              >
                Build a Pinterest Clone
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr>
            <td
              colSpan={2}
            >
              <a
                className="col-xs-12 btn btn-primary btn-block"
                data-cy="btn-for-legacy-back-end"
                href="/certification/undefined/legacy-back-end"
                onClick={[Function]}
                onKeyDown={[Function]}
              >
                buttons.claim-cert
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-sm-8 col-sm-offset-2 col-xs-12"
    >
      <div
        className="spacer"
        style={
          Object {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <h3
        className="text-center"
        id="cert-legacy-data-visualization"
      >
        Legacy Data Visualization
      </h3>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              settings.labels.project-name
            </th>
            <th>
              settings.labels.solution
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/front-end-development-libraries/front-end-development-libraries-projects/build-a-markdown-previewer"
              >
                Build a Markdown Previewer
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-freecodecamp-forum-homepage"
              >
                Build a freeCodeCamp Forum Homepage
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-recipe-box"
              >
                Build a Recipe Box
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-the-game-of-life"
              >
                Build the Game of Life
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/build-a-roguelike-dungeon-crawler-game"
              >
                Build a Roguelike Dungeon Crawler Game
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/data-visualization/data-visualization-projects/visualize-data-with-a-bar-chart"
              >
                Visualize Data with a Bar Chart
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/data-visualization/data-visualization-projects/visualize-data-with-a-scatterplot-graph"
              >
                Visualize Data with a Scatterplot Graph
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/data-visualization/data-visualization-projects/visualize-data-with-a-heat-map"
              >
                Visualize Data with a Heat Map
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/show-national-contiguity-with-a-force-directed-graph"
              >
                Show National Contiguity with a Force Directed Graph
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/coding-interview-prep/take-home-projects/map-data-across-the-globe"
              >
                Map Data Across the Globe
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr>
            <td
              colSpan={2}
            >
              <a
                className="col-xs-12 btn btn-primary btn-block"
                data-cy="btn-for-legacy-data-visualization"
                href="/certification/undefined/legacy-data-visualization"
                onClick={[Function]}
                onKeyDown={[Function]}
              >
                buttons.claim-cert
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-sm-8 col-sm-offset-2 col-xs-12"
    >
      <div
        className="spacer"
        style={
          Object {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <h3
        className="text-center"
        id="cert-information-security-and-quality-assurance"
      >
        Legacy Information Security and Quality Assurance
      </h3>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              settings.labels.project-name
            </th>
            <th>
              settings.labels.solution
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/quality-assurance/quality-assurance-projects/metric-imperial-converter"
              >
                Metric-Imperial Converter
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/quality-assurance/quality-assurance-projects/issue-tracker"
              >
                Issue Tracker
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/quality-assurance/quality-assurance-projects/personal-library"
              >
                Personal Library
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/information-security/information-security-projects/stock-price-checker"
              >
                Stock Price Checker
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr
            className="project-row"
          >
            <td
              className="project-title col-sm-8 col-xs-8"
            >
              <a
                href="/learn/information-security/information-security-projects/anonymous-message-board"
              >
                Anonymous Message Board
              </a>
            </td>
            <td
              className="project-solution col-sm-4 col-xs-4"
            />
          </tr>
          <tr>
            <td
              colSpan={2}
            >
              <a
                className="col-xs-12 btn btn-primary btn-block"
                data-cy="btn-for-information-security-and-quality-assurance"
                href="/certification/undefined/information-security-and-quality-assurance"
                onClick={[Function]}
                onKeyDown={[Function]}
              >
                buttons.claim-cert
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
]
`;
